core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -498,6 +498,45 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
       requestLocal)
   }
 
+  /*
+    The state transition for Empty, CompleteAbort, CompleteCommit is complicated, so here is a table to make it clear.
+    Note:
+    PF = PRODUCER_FENCED
+    ITS = INVALID_TXN_STATE
+    NONE = No error and no epoch bump
+    EB = No error and epoch bump
+
+    Retry => producerEpoch = txnState.ProducerEpoch - 1
```

Comment:
Hmm -- retry is a little interesting here, since the retry is only a retry if the state is prepare/complete abort? for the other cases, it is a fenced request? I think for v2 empty + "retry" should be its? Since if we just aborted, we can't go back to empty.

Reply:
I think that one was the only transition that didn't make sense to me. Thanks for enumerating it out.

Reply:
Makes sense, I agree. The empty state does not come from an abort, then there is no such concept of "Retry"